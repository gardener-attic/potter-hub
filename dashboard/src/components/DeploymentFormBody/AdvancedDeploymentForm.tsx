import * as React from "react";
import AceEditor from "react-ace";

import "ace-builds/src-noconflict/mode-yaml";
import "ace-builds/src-noconflict/theme-xcode";

export interface IAdvancedDeploymentForm {
  appValues?: string;
  handleValuesChange: (value: string) => void;
}

class AdvancedDeploymentForm extends React.Component<IAdvancedDeploymentForm> {
  public render() {
    return (
      <div className="margin-t-normal">
        <AceEditor
          mode="yaml"
          theme="xcode"
          width="100%"
          onChange={this.props.handleValuesChange}
          setOptions={{ showPrintMargin: false }}
          editorProps={{ $blockScrolling: Infinity }}
          value={this.props.appValues}
          className="editor"
        />
        <p>
          <b>Note:</b> Only comments from the original chart values will be preserved.
        </p>
      </div>
    );
  }
}

export default AdvancedDeploymentForm;
