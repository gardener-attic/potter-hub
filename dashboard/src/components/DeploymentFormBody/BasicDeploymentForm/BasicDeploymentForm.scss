.description {
  color: #5f6369;
  font-size: 0.9em;
}

.subsection {
  border: 1px solid #aaa;
  border-radius: 5px;
  padding: 10px;
}

.react-switch {
  vertical-align: middle;
  margin-left: 10px;
}

.disk_size_input {
  width: 75%;
}

.block:before {
  content: "";
  display: inline-block;
  height: 100%;
  vertical-align: middle;
}

.centered {
  display: inline-block;
}
