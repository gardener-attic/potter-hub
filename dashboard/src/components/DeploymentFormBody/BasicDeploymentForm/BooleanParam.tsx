import * as React from "react";
import Switch from "react-switch";
import { IBasicFormParam } from "shared/types";

export interface IStringParamProps {
  id: string;
  label: string;
  param: IBasicFormParam;
  handleBasicFormParamChange: (
    p: IBasicFormParam,
  ) => (e: React.FormEvent<HTMLInputElement>) => void;
}

class BooleanParam extends React.Component<IStringParamProps> {
  public render() {
    const { id, param, label } = this.props;
    return (
      <label htmlFor={id}>
        <div className="margin-b-normal">
          <span>{label}</span>
          <Switch
            id={id}
            onChange={this.handleChange}
            checked={!!param.value}
            className="react-switch"
          />
        </div>
        {param.description && <span className="description">{param.description}</span>}
      </label>
    );
  }

  // handleChange transform the event received by the Switch component to a checkbox event
  public handleChange = (checked: boolean) => {
    const { param } = this.props;
    const event = {
      currentTarget: { value: String(checked), type: "checkbox", checked },
    } as React.FormEvent<HTMLInputElement>;
    this.props.handleBasicFormParamChange(param)(event);
  };
}

export default BooleanParam;
