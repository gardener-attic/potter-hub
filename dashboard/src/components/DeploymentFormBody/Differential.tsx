import * as jsdiff from "diff";
import { Diff2Html } from "diff2html";
import * as React from "react";

import "diff2html/dist/diff2html.css";

export interface IDifferentialProps {
  title: string;
  oldValues: string;
  newValues: string;
  emptyDiffText: string;
}

class Differential extends React.Component<IDifferentialProps> {
  public render = () => {
    const { oldValues, newValues, title, emptyDiffText } = this.props;
    const sdiff = jsdiff.createPatch(title, oldValues, newValues);
    const outputHtml = Diff2Html.getPrettyHtml(sdiff, {
      inputFormat: "diff",
      showFiles: false,
      matching: "lines",
      maxLineSizeInBlockForComparison: 20,
    });
    return (
      <div className="diff">
        {oldValues === newValues ? (
          <p>{emptyDiffText}</p>
        ) : (
          <div dangerouslySetInnerHTML={{ __html: outputHtml }} />
        )}
      </div>
    );
  };
}

export default Differential;
