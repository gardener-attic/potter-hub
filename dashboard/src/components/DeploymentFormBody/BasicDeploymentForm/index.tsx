import BasicDeploymentForm from "./BasicDeploymentForm";

export default BasicDeploymentForm;
