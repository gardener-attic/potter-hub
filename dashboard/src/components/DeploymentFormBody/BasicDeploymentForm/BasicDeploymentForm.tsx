import { isArray } from "lodash";
import * as React from "react";
import { IBasicFormParam, IBasicFormSliderParam } from "shared/types";
import TextParam from "./TextParam";

import { getValue } from "../../../shared/schema";
import "./BasicDeploymentForm.css";
import BooleanParam from "./BooleanParam";
import SliderParam from "./SliderParam";
import Subsection from "./Subsection";

export interface IBasicDeploymentFormProps {
  params: IBasicFormParam[];
  handleBasicFormParamChange: (
    p: IBasicFormParam,
  ) => (e: React.FormEvent<HTMLInputElement | HTMLTextAreaElement | HTMLSelectElement>) => void;
  handleValuesChange: (value: string) => void;
  appValues: string;
}

class BasicDeploymentForm extends React.Component<IBasicDeploymentFormProps> {
  public render() {
    return (
      <div className="margin-t-normal">
        {this.props.params.map((param, i) => {
          const id = `${param.path}-${i}`;
          return (
            <div key={id}>
              {this.renderParam(param, id, this.props.handleBasicFormParamChange)}
              <hr />
            </div>
          );
        })}
      </div>
    );
  }

  private isHidden = (param: IBasicFormParam) => {
    const hidden = param.hidden;
    switch (typeof hidden) {
      case "string":
        // If hidden is a string, it points to the value that should be true
        return getValue(this.props.appValues, hidden) === true;
      case "object":
        // Two type of supported objects
        // A single condition: {value: string, path: any}
        // An array of conditions: {conditions: Array<{value: string, path: any}, operator: string}
        if (hidden.conditions?.length > 0) {
          // If hidden is an object, a different logic should be applied
          // based on the operator
          switch (hidden.operator) {
            case "and":
              // Every value matches the referenced
              // value (via jsonpath) in all the conditions
              return hidden.conditions.every(
                c => getValue(this.props.appValues, c.path) === c.value,
              );
            case "or":
              // It is enough if the value matches the referenced
              // value (via jsonpath) in any of the conditions
              return hidden.conditions.some(
                c => getValue(this.props.appValues, c.path) === c.value,
              );
            case "nor":
              // Every value mismatches the referenced
              // value (via jsonpath) in any of the conditions
              return hidden.conditions.every(
                c => getValue(this.props.appValues, c.path) !== c.value,
              );
            default:
              // we consider 'and' as the default operator
              return hidden.conditions.every(
                c => getValue(this.props.appValues, c.path) === c.value,
              );
          }
        } else {
          return getValue(this.props.appValues, hidden.path) === hidden.value;
        }
      case "undefined":
        return false;
    }
  };

  private renderParam = (
    param: IBasicFormParam,
    id: string,
    handleBasicFormParamChange: (
      p: IBasicFormParam,
    ) => (e: React.FormEvent<HTMLInputElement | HTMLTextAreaElement | HTMLSelectElement>) => void,
  ) => {
    let paramComponent: JSX.Element = <></>;
    // If the type of the param is an array, represent it as its first type
    const type = isArray(param.type) ? param.type[0] : param.type;
    switch (type) {
      case "boolean":
        paramComponent = (
          <BooleanParam
            label={param.title || param.path}
            handleBasicFormParamChange={handleBasicFormParamChange}
            id={id}
            param={param}
          />
        );
        break;
      case "object": {
        paramComponent = (
          <Subsection
            label={param.title || param.path}
            handleValuesChange={this.props.handleValuesChange}
            appValues={this.props.appValues}
            renderParam={this.renderParam}
            param={param}
          />
        );
        break;
      }
      case "string": {
        if (param.render === "slider") {
          const p = param as IBasicFormSliderParam;
          paramComponent = (
            <SliderParam
              label={param.title || param.path}
              handleBasicFormParamChange={handleBasicFormParamChange}
              id={id}
              param={param}
              min={p.sliderMin || 1}
              max={p.sliderMax || 1000}
              unit={p.sliderUnit || ""}
            />
          );
          break;
        }
        if (param.render === "textArea") {
          paramComponent = (
            <TextParam
              label={param.title || param.path}
              handleBasicFormParamChange={handleBasicFormParamChange}
              id={id}
              param={param}
              inputType="textarea"
            />
          );
          break;
        }
      }
      default:
        paramComponent = (
          <TextParam
            label={param.title || param.path}
            handleBasicFormParamChange={handleBasicFormParamChange}
            id={id}
            param={param}
            inputType={type === "integer" ? "number" : "string"}
          />
        );
    }
    return (
      <div key={id} hidden={this.isHidden(param)}>
        {paramComponent}
      </div>
    );
  };
}

export default BasicDeploymentForm;
