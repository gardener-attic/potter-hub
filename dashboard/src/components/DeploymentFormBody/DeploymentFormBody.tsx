import { RouterAction } from "connected-react-router";
import * as jsonpatch from "fast-json-patch";
import * as React from "react";
import { AlertOctagon } from "react-feather";
import { Tab, TabList, TabPanel, Tabs } from "react-tabs";

import { definedNamespaces } from "../../shared/Namespace";
import { retrieveBasicFormParams, setValue } from "../../shared/schema";
import { IBasicFormParam, IChartState } from "../../shared/types";
import { getValueFromEvent } from "../../shared/utils";
import ConfirmDialog from "../ConfirmDialog";
import { ErrorSelector } from "../ErrorAlert";
import Hint from "../Hint";
import LoadingWrapper from "../LoadingWrapper";
import AdvancedDeploymentForm from "./AdvancedDeploymentForm";
import BasicDeploymentForm from "./BasicDeploymentForm";

import "react-tabs/style/react-tabs.css";
import Differential from "./Differential";
import "./Tabs.css";

export interface IDeploymentFormBodyProps {
  chartID: string;
  chartVersion: string;
  deployedValues?: string;
  namespace: string;
  releaseVersion?: string;
  selected: IChartState["selected"];
  appValues: string;
  push: (location: string) => RouterAction;
  goBack?: () => RouterAction;
  getChartVersion: (id: string, chartVersion: string) => void;
  setValues: (values: string) => void;
  setValuesModified: () => void;
}

export interface IDeploymentFormBodyState {
  basicFormParameters: IBasicFormParam[];
  restoreDefaultValuesModalIsOpen: boolean;
  modifications?: jsonpatch.Operation[];
}

class DeploymentFormBody extends React.Component<
  IDeploymentFormBodyProps,
  IDeploymentFormBodyState
> {
  public state: IDeploymentFormBodyState = {
    basicFormParameters: [],
    restoreDefaultValuesModalIsOpen: false,
  };

  public componentDidMount() {
    const { chartID, getChartVersion, chartVersion } = this.props;
    getChartVersion(chartID, chartVersion);
  }

  public componentDidUpdate = (prevProps: IDeploymentFormBodyProps) => {
    const { chartID, chartVersion, getChartVersion, selected, appValues } = this.props;

    if (chartVersion !== prevProps.chartVersion) {
      // New version detected
      getChartVersion(chartID, chartVersion);
      return;
    }

    if (
      // If the selected schema changes
      prevProps.selected.schema !== selected.schema ||
      // or the selected values (because it's a different version)
      prevProps.selected.values !== selected.values ||
      // or the current values (and we hadn't process those values yet)
      (prevProps.appValues !== appValues && prevProps.appValues.length === 0)
    ) {
      // Parse the basic parameters
      this.setState({
        basicFormParameters: retrieveBasicFormParams(appValues, selected.schema),
      });
    }
  };

  public render() {
    const { selected, chartID, chartVersion, goBack, namespace } = this.props;
    const { version, versions } = selected;
    if (selected.error) {
      return (
        <ErrorSelector error={selected.error} resource={`Chart "${chartID}" (${chartVersion})`} />
      );
    }
    if (!version || !versions.length) {
      return <LoadingWrapper />;
    }
    return (
      <div>
        <ConfirmDialog
          modalIsOpen={this.state.restoreDefaultValuesModalIsOpen}
          loading={false}
          confirmationText={"Are you sure you want to restore the default chart values?"}
          confirmationButtonText={"Restore"}
          onConfirm={this.restoreDefaultValues}
          closeModal={this.closeRestoreDefaultValuesModal}
        />
        <div>
          <label htmlFor="chartVersion">Version</label>
          <select
            id="chartVersion"
            onChange={this.handleChartVersionChange}
            value={version.attributes.version}
            required={true}
          >
            {versions.map(v => (
              <option key={v.id} value={v.attributes.version}>
                {`Chart: ${v.attributes.version} | App: ${v.attributes.app_version || "-"} `} 
                {this.props.releaseVersion && v.attributes.version === this.props.releaseVersion
                  ? "(current)"
                  : ""}
              </option>
            ))}
          </select>
        </div>
        {this.renderTabs()}
        <div className="margin-t-big">
          <button
            className="button button-primary"
            type="submit"
            disabled={!namespace || namespace === definedNamespaces.all}
          >
            Submit
          </button>
          <button className="button" type="button" onClick={this.openRestoreDefaultValuesModal}>
            Restore Chart Defaults
          </button>
          {goBack && (
            <button className="button" type="button" onClick={goBack}>
              Back
            </button>
          )}
          {(!namespace || namespace === definedNamespaces.all) && (
            <div className="container-tiny" style={{ display: "flex", marginTop: "10px" }}>
              <div className="alert alert-warning AlertBox" role="alert">
                <div className="AlertBoxIconContainer">
                  <AlertOctagon />
                </div>
                Please select a namespace using the namespace selector from the top menu.
              </div>
            </div>
          )}
        </div>
      </div>
    );
  }

  private handleChartVersionChange = (e: React.FormEvent<HTMLSelectElement>) => {
    // TODO(andres): This requires refactoring. Currently, the deploy and upgrade
    // forms behave differently. In the deployment form, a change in the version
    // changes the route but in the case of the upgrade it only changes the state
    const isUpgradeForm = !!this.props.releaseVersion;

    if (isUpgradeForm) {
      const { chartID, getChartVersion } = this.props;
      getChartVersion(chartID, e.currentTarget.value);
    } else {
      this.props.push(
        `/apps/ns/${this.props.namespace}/new/${this.props.chartID}/versions/${e.currentTarget.value}`,
      );
    }
  };

  private handleValuesChange = (value: string) => {
    this.props.setValues(value);
    this.props.setValuesModified();
  };

  private refreshBasicParameters = () => {
    this.setState({
      basicFormParameters: retrieveBasicFormParams(
        this.props.appValues,
        this.props.selected.schema,
      ),
    });
  };

  private renderTabs = () => {
    return (
      <div className="margin-t-normal">
        <Tabs>
          <TabList>
            {this.shouldRenderBasicForm() && (
              <Tab onClick={this.refreshBasicParameters}>
                Form{" "}
                <Hint reactTooltipOpts={{ delayHide: 100 }} id="basicFormHelp">
                  <span>
                    This form has been automatically generated based on the chart schema.
                    <br />
                    This feature is currently in a beta state. If you find an issue please report it{" "}
                    <a
                      target="_blank"
                      rel="noopener noreferrer"
                      href="https://github.com/gardener/potter-hub/issues/new/choose"
                    >
                      here.
                    </a>
                  </span>
                </Hint>
              </Tab>
            )}
            <Tab>Values (YAML)</Tab>
            <Tab>Changes</Tab>
          </TabList>
          {this.shouldRenderBasicForm() && (
            <TabPanel>
              <BasicDeploymentForm
                params={this.state.basicFormParameters}
                handleBasicFormParamChange={this.handleBasicFormParamChange}
                appValues={this.props.appValues}
                handleValuesChange={this.handleValuesChange}
              />
            </TabPanel>
          )}
          <TabPanel>
            <AdvancedDeploymentForm
              appValues={this.props.appValues}
              handleValuesChange={this.handleValuesChange}
            />
          </TabPanel>
          <TabPanel>{this.renderDiff()}</TabPanel>
        </Tabs>
      </div>
    );
  };

  private handleBasicFormParamChange = (param: IBasicFormParam) => {
    return (e: React.FormEvent<any>) => {
      this.props.setValuesModified();
      const value = getValueFromEvent(e);
      this.setState({
        // Change param definition
        basicFormParameters: this.state.basicFormParameters.map(p =>
          p.path === param.path ? { ...param, value } : p,
        ),
      });
      // Change raw values
      this.props.setValues(setValue(this.props.appValues, param.path, value));
    };
  };

  // The basic form should be rendered if there are params to show
  private shouldRenderBasicForm = () => {
    return Object.keys(this.state.basicFormParameters).length > 0;
  };

  private closeRestoreDefaultValuesModal = () => {
    this.setState({ restoreDefaultValuesModalIsOpen: false });
  };

  private openRestoreDefaultValuesModal = () => {
    this.setState({ restoreDefaultValuesModalIsOpen: true });
  };

  private restoreDefaultValues = () => {
    if (this.props.selected.values) {
      this.props.setValues(this.props.selected.values);
      this.setState({
        basicFormParameters: retrieveBasicFormParams(
          this.props.selected.values,
          this.props.selected.schema,
        ),
      });
    }
    this.setState({ restoreDefaultValuesModalIsOpen: false });
  };

  private renderDiff = () => {
    let oldValues = "";
    let title = "";
    let emptyDiffText = "";
    if (this.props.deployedValues) {
      // If there are already some deployed values (upgrade scenario)
      // We compare the values from the old release and the new one
      oldValues = this.props.deployedValues;
      title = "Difference from deployed version";
      emptyDiffText = "The values for the new release are identical to the deployed version.";
    } else {
      // If it's a new deployment, we show the different from the default
      // values for the selected version
      oldValues = this.props.selected.values || "";
      title = "Difference from chart defaults";
      emptyDiffText = "No changes detected from chart defaults.";
    }
    return (
      <Differential
        title={title}
        oldValues={oldValues}
        newValues={this.props.appValues}
        emptyDiffText={emptyDiffText}
      />
    );
  };
}

export default DeploymentFormBody;
