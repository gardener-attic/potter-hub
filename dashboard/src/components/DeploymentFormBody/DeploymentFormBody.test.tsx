import { mount, shallow } from "enzyme";
import * as React from "react";

import { Tab, Tabs } from "react-tabs";
import itBehavesLike from "../../shared/specs";
import { IChartState, IChartVersion, NotFoundError } from "../../shared/types";
import ConfirmDialog from "../ConfirmDialog";
import { ErrorSelector } from "../ErrorAlert";
import ErrorPageHeader from "../ErrorAlert/ErrorAlertHeader";
import LoadingWrapper from "../LoadingWrapper";
import BasicDeploymentForm from "./BasicDeploymentForm/BasicDeploymentForm";
import DeploymentFormBody, {
  IDeploymentFormBodyProps,
  IDeploymentFormBodyState,
} from "./DeploymentFormBody";
import Differential from "./Differential";

const defaultProps = {
  chartID: "foo",
  chartVersion: "1.0.0",
  error: undefined,
  releaseName: undefined,
  selected: {} as IChartState["selected"],
  deployChart: jest.fn(),
  push: jest.fn(),
  fetchChartVersions: jest.fn(),
  getChartVersion: jest.fn(),
  namespace: "default",
  appValues: "foo: bar",
  valuesModified: false,
  setValues: jest.fn(),
  setValuesModified: jest.fn(),
} as IDeploymentFormBodyProps;
const versions = [{ id: "foo", attributes: { version: "1.2.3", app_version: "4.5.6" } }] as IChartVersion[];

itBehavesLike("aLoadingComponent", { component: DeploymentFormBody, props: defaultProps });

describe("renders an error", () => {
  it("renders an error if it cannot find the given chart", () => {
    const wrapper = mount(
      <DeploymentFormBody
        {...defaultProps}
        selected={{ error: new NotFoundError() } as IChartState["selected"]}
      />,
    );
    expect(wrapper.find(ErrorPageHeader).exists()).toBe(true);
    expect(wrapper.find(ErrorPageHeader).text()).toContain('Chart "foo" (1.0.0) not found');
  });

  it("renders a generic error", () => {
    const wrapper = shallow(
      <DeploymentFormBody
        {...defaultProps}
        selected={{ error: new Error() } as IChartState["selected"]}
      />,
    );
    expect(wrapper.find(ErrorSelector).exists()).toBe(true);
    expect(wrapper.find(ErrorSelector).html()).toContain("Sorry! Something went wrong");
  });
});

it("renders the full DeploymentFormBody", () => {
  const wrapper = shallow(
    <DeploymentFormBody {...defaultProps} selected={{ versions, version: versions[0] }} />,
  );
  expect(wrapper).toMatchSnapshot();
});

it("marks the current version", () => {
  const wrapper = shallow(
    <DeploymentFormBody
      {...defaultProps}
      releaseVersion={versions[0].attributes.version}
      selected={{ versions, version: versions[0] }}
    />,
  );
  expect(wrapper.find("select").text()).toMatch("Chart: 1.2.3 | App: 4.5.6 (current)");
});

const initialValues = "foo: bar";
const initialSchema = { properties: { foo: { type: "string", form: true } } };
const chartVersion = {
  id: "foo",
  attributes: { version: "1.0.0", app_version: "1.0", created: "1" },
  relationships: {
    chart: {
      data: {
        name: "chart",
        description: "chart-description",
        keywords: [],
        maintainers: [],
        repo: {
          name: "repo",
          url: "http://example.com",
        },
        sources: [],
      },
    },
  },
};
const props: IDeploymentFormBodyProps = {
  ...defaultProps,
  selected: {
    ...defaultProps.selected,
    versions: [chartVersion],
    version: chartVersion,
    values: initialValues,
    schema: initialSchema,
  },
};

describe("when there are changes in the selected version", () => {
  it("initializes the local values from props when props set", () => {
    const wrapper = shallow(<DeploymentFormBody {...defaultProps} />);
    wrapper.setProps({ selected: props.selected });
    const localState: IDeploymentFormBodyState = wrapper.instance()
      .state as IDeploymentFormBodyState;
    const basicFormParameters = [
      {
        form: true,
        path: "foo",
        value: "bar",
        type: "string",
      },
    ];
    expect(localState.basicFormParameters).toEqual(basicFormParameters);
  });
});

describe("when the basic form is enabled", () => {
  it("renders the different tabs", () => {
    const basicFormParameters = [
      {
        path: "wordpressUsername",
        value: "user",
      },
    ];
    const wrapper = mount(<DeploymentFormBody {...props} />);
    wrapper.setState({ appValues: "wordpressUsername: user", basicFormParameters });
    wrapper.update();
    expect(wrapper.find(LoadingWrapper)).not.toExist();
    expect(wrapper.find(Tabs)).toExist();
  });

  it("changes the parameter value", () => {
    const basicFormParameters = [
      {
        path: "wordpressUsername",
        value: "user",
      },
    ];
    const setValuesModified = jest.fn();
    const setValues = jest.fn();
    const wrapper = mount(
      <DeploymentFormBody
        {...props}
        appValues="wordpressUsername: user"
        setValues={setValues}
        setValuesModified={setValuesModified}
      />,
    );
    wrapper.setState({ basicFormParameters });
    wrapper.update();

    // Fake onChange
    const input = wrapper.find(BasicDeploymentForm).find("input");
    const onChange = input.prop("onChange") as (e: React.FormEvent<HTMLInputElement>) => void;
    onChange({ currentTarget: { value: "foo" } } as React.FormEvent<HTMLInputElement>);

    expect(wrapper.state("basicFormParameters")).toEqual([
      { path: "wordpressUsername", value: "foo" },
    ]);
    expect(setValuesModified).toHaveBeenCalled();
    expect(setValues).toHaveBeenCalledWith("wordpressUsername: foo\n");
  });

  it("should update existing params if the app values change and the user clicks on the Basic tab", () => {
    const testProps = {
      ...props,
      selected: {
        ...props.selected,
        schema: { properties: { wordpressUsername: { type: "string", form: true } } },
      },
    };
    const basicFormParameters = [
      {
        path: "wordpressUsername",
        value: "user",
      },
    ];
    const wrapper = mount(<DeploymentFormBody {...testProps} />);
    wrapper.setState({ basicFormParameters });
    wrapper.setProps({ appValues: "wordpressUsername: foo" });
    wrapper.update();

    const tab = wrapper
      .find(Tab)
      .findWhere(t => !!t.text().match("Form"))
      .first();
    tab.simulate("click");

    expect(wrapper.state("basicFormParameters")).toMatchObject([
      {
        path: "wordpressUsername",
        value: "foo",
      },
    ]);
  });

  it("should update existing params when receiving new values", () => {
    const testProps = {
      ...props,
      selected: {
        ...props.selected,
        schema: { properties: { wordpressUsername: { type: "string", form: true } } },
      },
      appValues: "",
    };
    const basicFormParameters = [
      {
        path: "wordpressUsername",
        value: "user",
      },
    ];
    const wrapper = mount(<DeploymentFormBody {...testProps} />);
    wrapper.setState({ basicFormParameters });
    wrapper.setProps({ appValues: "wordpressUsername: foo" });
    wrapper.update();
    expect(wrapper.state("basicFormParameters")).toMatchObject([
      {
        path: "wordpressUsername",
        value: "foo",
      },
    ]);
  });

  it("should update existing params when receiving new values (for a new version)", () => {
    const testProps = {
      ...props,
      selected: {
        ...props.selected,
        schema: { properties: { wordpressUsername: { type: "string", form: true } } },
        values: "wordpressUsername: foo",
      },
    };
    const basicFormParameters = [
      {
        path: "wordpressUsername",
        value: "user",
      },
    ];
    const wrapper = mount(<DeploymentFormBody {...testProps} />);
    wrapper.setState({ basicFormParameters });
    const updatedProps = {
      selected: {
        ...props.selected,
        schema: { properties: { wordpressUsername: { type: "string", form: true } } },
        values: "wordpressUsername: bar",
      },
      appValues: "wordpressUsername: bar",
    };

    wrapper.setProps(updatedProps);
    wrapper.update();
    expect(wrapper.state("basicFormParameters")).toMatchObject([
      {
        path: "wordpressUsername",
        value: "bar",
      },
    ]);
  });

  it("should not re-render the basic params if appValues changes (because it's handled by the parameter itself)", () => {
    const testProps = {
      ...props,
      selected: {
        ...props.selected,
        schema: { properties: { wordpressUsername: { type: "string", form: true } } },
      },
      appValues: "wordpressUsername: foo",
    };
    const basicFormParameters = [
      {
        path: "wordpressUsername",
        value: "foo",
      },
    ];
    const wrapper = mount(<DeploymentFormBody {...testProps} />);
    wrapper.setState({ basicFormParameters });
    wrapper.setProps({ appValues: "wordpressUsername: bar" });
    wrapper.update();
    expect(wrapper.state("basicFormParameters")).toMatchObject([
      {
        path: "wordpressUsername",
        value: "foo",
      },
    ]);
  });

  it("handles a parameter as a number", () => {
    const setValues = jest.fn();
    const testProps = {
      ...props,
      selected: {
        ...props.selected,
        schema: { properties: { replicas: { type: "integer", form: true } } },
      },
    };
    const basicFormParameters = [
      {
        form: true,
        path: "replicas",
        value: 1,
        type: "integer",
      },
    ];
    const wrapper = mount(<DeploymentFormBody {...testProps} setValues={setValues} />);
    wrapper.setState({ basicFormParameters });
    wrapper.setProps({ appValues: "replicas: 1" });
    wrapper.update();

    // Fake onChange
    const input = wrapper.find(BasicDeploymentForm).find("input");
    const onChange = input.prop("onChange") as (e: React.FormEvent<HTMLInputElement>) => void;
    onChange({ currentTarget: { value: "2", valueAsNumber: 2, type: "number" } } as React.FormEvent<
      HTMLInputElement
    >);

    expect(wrapper.state("basicFormParameters")).toEqual([
      {
        form: true,
        path: "replicas",
        value: 2,
        type: "integer",
      },
    ]);
    expect(setValues).toHaveBeenCalledWith("replicas: 2\n");
  });

  it("handles a parameter as a boolean", () => {
    const testProps = {
      ...props,
      selected: {
        ...props.selected,
        schema: { properties: { enableMetrics: { type: "boolean", form: true } } },
      },
    };
    const setValues = jest.fn();
    const basicFormParameters = [
      {
        form: true,
        path: "enableMetrics",
        value: false,
        type: "boolean",
      },
    ];
    const wrapper = mount(<DeploymentFormBody {...testProps} setValues={setValues} />);
    wrapper.setState({ basicFormParameters });
    wrapper.setProps({ appValues: "enableMetrics: false" });
    wrapper.update();

    // Fake onChange
    const input = wrapper.find(BasicDeploymentForm).find("input");
    const onChange = input.prop("onChange") as (e: React.FormEvent<HTMLInputElement>) => void;
    onChange({
      currentTarget: { value: "true", checked: true, type: "checkbox" },
    } as React.FormEvent<HTMLInputElement>);

    expect(wrapper.state("basicFormParameters")).toEqual([
      {
        form: true,
        path: "enableMetrics",
        value: true,
        type: "boolean",
      },
    ]);
    expect(setValues).toHaveBeenCalledWith("enableMetrics: true\n");
  });
});

it("goes back when clicking in the Back button", () => {
  const goBack = jest.fn();
  const wrapper = shallow(<DeploymentFormBody {...props} goBack={goBack} />);
  const backButton = wrapper.find(".button").filterWhere(i => i.text() === "Back");
  expect(backButton).toExist();
  // Avoid empty or submit type
  expect(backButton.prop("type")).toBe("button");
  backButton.simulate("click");
  expect(goBack).toBeCalled();
});

it("restores the default chart values when clicking on the button", () => {
  const setValues = jest.fn();
  const wrapper = shallow(
    <DeploymentFormBody
      {...props}
      setValues={setValues}
      selected={{
        ...props.selected,
        values: "foo: value",
      }}
    />,
  );

  // bypass modal
  wrapper.find(ConfirmDialog).prop("onConfirm")();

  expect(setValues).toHaveBeenCalledWith("foo: value");
});

describe("Changes tab", () => {
  it("should show the differences between the default chart values when deploying", () => {
    const selected = {
      ...defaultProps.selected,
      versions: [chartVersion],
      version: chartVersion,
      values: "foo: bar",
      schema: initialSchema,
    };
    const appValues = "bar: foo";
    const wrapper = shallow(
      <DeploymentFormBody {...props} selected={selected} appValues={appValues} />,
    );

    const Diff = wrapper.find(Differential);
    expect(Diff.props()).toMatchObject({
      emptyDiffText: "No changes detected from chart defaults.",
      newValues: "bar: foo",
      oldValues: "foo: bar",
      title: "Difference from chart defaults",
    });
  });

  it("should show the differences between the current release and the new one when upgrading", () => {
    const selected = {
      ...defaultProps.selected,
      versions: [chartVersion],
      version: chartVersion,
      values: "foo: bar",
      schema: initialSchema,
    };
    const deployedValues = "a: b";
    const appValues = "bar: foo";
    const wrapper = shallow(
      <DeploymentFormBody
        {...props}
        selected={selected}
        appValues={appValues}
        deployedValues={deployedValues}
      />,
    );

    const Diff = wrapper.find(Differential);
    expect(Diff.props()).toMatchObject({
      emptyDiffText: "The values for the new release are identical to the deployed version.",
      newValues: "bar: foo",
      oldValues: "a: b",
      title: "Difference from deployed version",
    });
  });
});
