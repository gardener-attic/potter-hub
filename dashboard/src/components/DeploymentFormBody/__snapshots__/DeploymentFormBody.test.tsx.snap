// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loading spinner matches the snapshot 1`] = `
<LoadingWrapper
  loaded={false}
  type={0}
/>
`;

exports[`renders the full DeploymentFormBody 1`] = `
<div>
  <ConfirmDialog
    closeModal={[Function]}
    confirmationButtonText="Restore"
    confirmationText="Are you sure you want to restore the default chart values?"
    loading={false}
    modalIsOpen={false}
    onConfirm={[Function]}
  />
  <div>
    <label
      htmlFor="chartVersion"
    >
      Version
    </label>
    <select
      id="chartVersion"
      onChange={[Function]}
      required={true}
      value="1.2.3"
    >
      <option
        key="foo"
        value="1.2.3"
      >
        Chart: 1.2.3 | App: 4.5.6 
      </option>
    </select>
  </div>
  <div
    className="margin-t-normal"
  >
    <Tabs
      defaultFocus={false}
      defaultIndex={null}
      forceRenderTabPanel={false}
      selectedIndex={null}
    >
      <TabList
        className="react-tabs__tab-list"
      >
        <Tab
          className="react-tabs__tab"
          disabledClassName="react-tabs__tab--disabled"
          focus={false}
          id={null}
          panelId={null}
          selected={false}
          selectedClassName="react-tabs__tab--selected"
        >
          Values (YAML)
        </Tab>
        <Tab
          className="react-tabs__tab"
          disabledClassName="react-tabs__tab--disabled"
          focus={false}
          id={null}
          panelId={null}
          selected={false}
          selectedClassName="react-tabs__tab--selected"
        >
          Changes
        </Tab>
      </TabList>
      <TabPanel
        className="react-tabs__tab-panel"
        forceRender={false}
        selectedClassName="react-tabs__tab-panel--selected"
      >
        <AdvancedDeploymentForm
          appValues="foo: bar"
          handleValuesChange={[Function]}
        />
      </TabPanel>
      <TabPanel
        className="react-tabs__tab-panel"
        forceRender={false}
        selectedClassName="react-tabs__tab-panel--selected"
      >
        <Differential
          emptyDiffText="No changes detected from chart defaults."
          newValues="foo: bar"
          oldValues=""
          title="Difference from chart defaults"
        />
      </TabPanel>
    </Tabs>
  </div>
  <div
    className="margin-t-big"
  >
    <button
      className="button button-primary"
      disabled={false}
      type="submit"
    >
      Submit
    </button>
    <button
      className="button"
      onClick={[Function]}
      type="button"
    >
      Restore Chart Defaults
    </button>
  </div>
</div>
`;
